### DISCUSSION TITLE
Asymmetry in shell element model with symmetric loading.

### DISCUSSION THREAD

**Discussion by Edward-Eth** [idaholab/moose#20771] [Q&A Modules: Solid mechanics] [answered]
I have a 2d shell element based model with full symmetry about the x axis, subject to a uniformly distributed pressure load on one edge, x displacement fixed on the other edge and y displacement fixed at the mid point of the edges. This model is fully symmetrical about the x axis. When run however, the results for stress in the 1(y) direction are asymmetric to a substantial degree (See image of contours). Weirdly however, the y displacements seem to be perfectly symmetrical. When I run a mesh that isn't tapered the y stress vanishes as you'd expect, and the X stress becomes uniform, so it seems that the issue is only arising when the mesh is tapered. I would appreciate any insight/advice that could be given on this as I've somewhat reached the end of my troubleshooting abilities in figuring this one out.

Stress 00 (xx) shows symmetry:
<img width="1112" alt="Screenshot 2022-04-12 at 14 49 32" src="https://user-images.githubusercontent.com/64844573/162979785-79debcb2-ce93-43df-8bc1-0e14f35a249d.png">

Displacement in x shows symmetry:
<img width="1112" alt="Screenshot 2022-04-12 at 14 50 40" src="https://user-images.githubusercontent.com/64844573/162979853-2aa202b3-5957-43b2-9701-69326b4f74f8.png">

Additionally, the displacement in y also seems to show symmetry:
<img width="1112" alt="Screenshot 2022-04-12 at 14 50 52" src="https://user-images.githubusercontent.com/64844573/162980647-47f43ba0-628f-44bb-b471-8dead5fa1321.png">

However stress in 01 (xy) and 11 (yy) directions show asymmetry:
<img width="1112" alt="Screenshot 2022-04-12 at 14 49 51" src="https://user-images.githubusercontent.com/64844573/162980007-371c6beb-70bb-4612-a60f-ac153b113051.png">
<img width="1112" alt="Screenshot 2022-04-12 at 14 49 43" src="https://user-images.githubusercontent.com/64844573/162980057-9188b44e-3214-4b2e-a5e4-557138156fe0.png">

The stress values in y are comparable in magnitude to the values in x (160 in yy direction to 1000 in xx) which means this isn't just a numerical error on a value that is vanishingly small.

The input file used:
```
[Mesh]
  [base]
    type = FileMeshGenerator
    file = ShellTest1.ucd
  []
  [side_sets]
    type = SideSetsFromNormalsGenerator
    input = base
    new_boundary = 'left right'
    normals = '-1 0 0
               1 0 0'
    fixed_normal = true
  []
  [left_side_mid_node]
    type = ExtraNodesetGenerator
    input = side_sets
    new_boundary = left_side_mid
    coord = '0.0 0.0 0.0'
  []
  [right_side_mid_node]
    type = ExtraNodesetGenerator
    input = left_side_mid_node
    new_boundary = right_side_mid
    coord = '4.0 0.0 0.0'
  []
[]

[Variables]
  [disp_x]
  []
  [disp_y]
  []
  [disp_z]
  []
  [rot_x]
  []
  [rot_y]
  []
[]

[AuxVariables]
  [stress_00]
  order = CONSTANT
  family = MONOMIAL
  []
  [stress_11]
  order = CONSTANT
  family = MONOMIAL
  []
  [stress_22]
  order = CONSTANT
  family = MONOMIAL
  []
  [stress_01]
  order = CONSTANT
  family = MONOMIAL
  []
  [stress_10]
  order = CONSTANT
  family = MONOMIAL
  []
  [stress_02]
  order = CONSTANT
  family = MONOMIAL
  []
  [stress_20]
  order = CONSTANT
  family = MONOMIAL
  []
  [stress_12]
  order = CONSTANT
  family = MONOMIAL
  []
  [stress_21]
  order = CONSTANT
  family = MONOMIAL
  []
[]

[Kernels]
  [solid_disp_x]
    type = ADStressDivergenceShell
    block = 1
    component = 0
    variable = disp_x
    through_thickness_order = SECOND
  []
  [solid_disp_y]
    type = ADStressDivergenceShell
    block = 1
    component = 1
    variable = disp_y
    through_thickness_order = SECOND
  []
  [solid_disp_z]
    type = ADStressDivergenceShell
    block = 1
    component = 2
    variable = disp_z
    through_thickness_order = SECOND
  []
  [solid_rot_x]
    type = ADStressDivergenceShell
    block = 1
    component = 3
    variable = rot_x
    through_thickness_order = SECOND
  []
  [solid_rot_y]
    type = ADStressDivergenceShell
    block = 1
    component = 4
    variable = rot_y
    through_thickness_order = SECOND
  []
[]

[AuxKernels]
  [stress_00]
    type = RankTwoAux
    variable = stress_00
    rank_two_tensor = global_stress_t_points_0
    index_i = 0
    index_j = 0
    execute_on = TIMESTEP_END
  []
  [stress_11]
    type = RankTwoAux
    variable = stress_11
    rank_two_tensor = global_stress_t_points_0
    index_i = 1
    index_j = 1
    execute_on = TIMESTEP_END
  []
  [stress_22]
    type = RankTwoAux
    variable = stress_22
    rank_two_tensor = global_stress_t_points_0
    index_i = 2
    index_j = 2
    execute_on = TIMESTEP_END
  []
  [stress_01]
    type = RankTwoAux
    variable = stress_01
    rank_two_tensor = global_stress_t_points_0
    index_i = 0
    index_j = 1
    execute_on = TIMESTEP_END
  []
  [stress_10]
    type = RankTwoAux
    variable = stress_10
    rank_two_tensor = global_stress_t_points_0
    index_i = 1
    index_j = 0
    execute_on = TIMESTEP_END
  []
  [stress_02]
    type = RankTwoAux
    variable = stress_02
    rank_two_tensor = global_stress_t_points_0
    index_i = 0
    index_j = 2
    execute_on = TIMESTEP_END
  []
  [stress_20]
    type = RankTwoAux
    variable = stress_20
    rank_two_tensor = global_stress_t_points_0
    index_i = 2
    index_j = 0
    execute_on = TIMESTEP_END
  []
  [stress_12]
    type = RankTwoAux
    variable = stress_12
    rank_two_tensor = global_stress_t_points_0
    index_i = 1
    index_j = 2
    execute_on = TIMESTEP_END
  []
  [stress_21]
    type = RankTwoAux
    variable = stress_21
    rank_two_tensor = global_stress_t_points_0
    index_i = 2
    index_j = 1
    execute_on = TIMESTEP_END
  []
[]

[BCs]
  [BC_0]
    type = ADDirichletBC
    variable = disp_x
    value = 0.0
    boundary = left
  []
  [BC_1]
    type = ADDirichletBC
    variable = disp_y
    value = 0.0
    boundary = left_side_mid
  []
  [BC_2]
    type = ADPressure
    variable = disp_x
    component = 0
    constant = -100.0
    boundary = right
    use_displaced_mesh = false
  []
[]

[Materials]
  [stress]
    type = ADComputeShellStress
    block = 1
    through_thickness_order = SECOND
  []
  [elasticity]
    type = ADComputeIsotropicElasticityTensorShell
    youngs_modulus = 210000
    poissons_ratio = 0.3
    block = 1
    through_thickness_order = SECOND
  []
  [strain]
    type = ADComputeIncrementalShellStrain
    block = 1
    displacements = 'disp_x disp_y disp_z'
    rotations = 'rot_x rot_y'
    thickness = 0.1
    through_thickness_order = SECOND
  []
[]

[Executioner]
  type = Steady
  solve_type = NEWTON
  automatic_scaling = true
  nl_abs_tol = 1e-30
  l_abs_tol = 1e-30
  nl_max_its = 1000
  l_max_its = 1000
[]

[Outputs]
  [outfile]
    type = Exodus
  []
[]
```

And the mesh file:
```
25 16 0 0 0 
1 0.0 2.0 0.0 
2 1.0 1.75 0.0 
3 2.0 1.5 0.0 
4 3.0 1.25 0.0 
5 4.0 1.0 0.0 
6 0.0 1.0 0.0 
7 1.0 0.875 0.0 
8 2.0 0.75 0.0 
9 3.0 0.625 0.0 
10 4.0 0.5 0.0 
11 0.0 0.0 0.0 
12 1.0 0.0 0.0 
13 2.0 0.0 0.0 
14 3.0 0.0 0.0 
15 4.0 0.0 0.0 
16 0.0 -1.0 0.0 
17 1.0 -0.875 0.0 
18 2.0 -0.75 0.0 
19 3.0 -0.625 0.0 
20 4.0 -0.5 0.0 
21 0.0 -2.0 0.0 
22 1.0 -1.75 0.0 
23 2.0 -1.5 0.0 
24 3.0 -1.25 0.0 
25 4.0 -1.0 0.0 
1 1 quad 1 6 7 2 
2 1 quad 2 7 8 3 
3 1 quad 3 8 9 4 
4 1 quad 4 9 10 5 
5 1 quad 6 11 12 7 
6 1 quad 7 12 13 8 
7 1 quad 8 13 14 9 
8 1 quad 9 14 15 10 
9 1 quad 11 16 17 12 
10 1 quad 12 17 18 13 
11 1 quad 13 18 19 14 
12 1 quad 14 19 20 15 
13 1 quad 16 21 22 17 
14 1 quad 17 22 23 18 
15 1 quad 18 23 24 19 
16 1 quad 19 24 25 20
```

**Comment by Edward-Eth** [comment]
I should also mention that mesh refinement doesn't resolve this issue:
<img width="1112" alt="Screenshot 2022-04-12 at 15 12 14" src="https://user-images.githubusercontent.com/64844573/162982049-a4e3d0dd-c1e8-4601-8461-e29700a9f6f2.png">
<img width="1112" alt="Screenshot 2022-04-12 at 15 12 20" src="https://user-images.githubusercontent.com/64844573/162982061-9689a3e2-7909-48c8-bf8b-5436a32e15ce.png">

**Comment by Edward-Eth** [comment]
Could someone at leas run this in their own moose-app to let me know if the issue is reproducible?

**Reply by GiudGiud** [reply]
@bwspenc @dschwen @jiangwen84 who works on shell elements?

**Reply by Edward-Eth** [reply]
Anyone had a chance to look at this yet?

**Reply by GiudGiud** [reply]
@jain651

**Reply by jain651** [reply]
@hrishiv would you please also look into this? Maybe create a test case along with others (that you are already developing to represent this problem.

**Reply by neuphris** [reply]
@Edward-Eth Thanks for pointing us to a bug in shell element. I was able to replicate it and figure out that the rotation matrix is calculated just once at the start of the simulation and is not stored. Hence, the rotation matrix for the last element was used to transform the stress and strain for all the elements giving incorrect stress and strain distribution. I
 was able to get the symmetric stress distribution by storing the rotation matrices. Will open a pull request for this.

![Screen Shot 2022-06-21 at 1 57 08 PM](https://user-images.githubusercontent.com/106320412/175130520-d3557193-1d71-481c-bb53-3368b0b954cf.png)
![Screen Shot 2022-06-21 at 1 56 24 PM](https://user-images.githubusercontent.com/106320412/175130560-f15205c7-9b16-4e97-bbee-0e3240fe47e6.png)

**Reply by Edward-Eth** [reply]
Awesome, glad I could be of assistance and thanks for fixing it 👍

**Comment by Edward-Eth** [comment, marked as answer]
Issue resolved pending Pull Request #21385 for the fixed code.

**Reply by neuphris** [reply]
Just wanted to give you heads up with applying the pressure BC to shell element. The pressure is designed to be applied over an area and when you are applying  pressure to the shell edge it uses only two quadrature points. Hence the team decided to generate an error message when pressure is applied to the edge as several MOOSE test cases were failing when a test file with pressure BC was used for shell element. You can use the resultant force to each node/edge instead of the pressure.

**Reply by Edward-Eth** [reply]
How would I set resultant force? Would that be a neumann BC or would it be best to use something like a ConstantRate nodal kernel?
